rclcpp/src/rclcpp/clock.cpp

```diff
@@ -182,6 +182,62 @@ Clock::sleep_for(Duration rel_time, Context::SharedPtr context)
   return sleep_until(now() + rel_time, context);
 }
 
+bool
+Clock::is_valid()
+{
+  switch (get_clock_type()) {
+    case RCL_ROS_TIME:
+    case RCL_STEADY_TIME:
+    case RCL_SYSTEM_TIME:
+      return now().nanoseconds() > 0;
+
+    case RCL_CLOCK_UNINITIALIZED:
+    default:
+      return false;
+  }
+}
+
+bool
+Clock::wait_for_valid(Context::SharedPtr context)
+{
+  if (get_clock_type() == RCL_CLOCK_UNINITIALIZED) {
+    throw std::runtime_error(
+            "clock cannot be waited on as its clock_type is RCL_CLOCK_UNINITIALIZED");
+  }
```

Comment:
What about the following?

```suggestion
  if (!is_valid()) {
    RCUTILS_LOG_ERROR("clock is not valid"); // this should be put in is_valid() method.
    return false;
  }
```

Reply:
This is gonna be a little confusing, but the `rcl_clock_valid` notion of validity (validity of the internal clock handle) is different from the notion of validity for time sources (as defined in the [design docs](https://design.ros2.org/articles/clock_and_time.html))

The rcl notion is concerned with being able to get a time from a clock. But the notion of validity being put forward here is for when a non-zero time is obtained from the clock.

So we can't just fail if a clock isn't `is_valid()`, since in some cases (when the clock has time 0 (and so is `rcl_clock_valid()` but not `is_valid()`), we can wait for the clock **to become** valid.

That said, it makes sense to put an RCUTILS_LOG_ERROR (or an RCUTILS_LOG_WARN, at least) if the clock will never be valid (in the case of a missing now() implementation or a clock uninitialization.)